allennlp/common/from_params.py

```diff
@@ -468,6 +467,31 @@ def constructor(**kwargs):
             return value_cls.from_params(params=deepcopy(popped_params), **constructor_extras)
 
         return Lazy(constructor)  # type: ignore
+
+    # This condition has to be at the end because other types can also be iterables
+    # (e.g., `str`, `tuple`, and `dict`).
+    elif (
+        origin in {collections.abc.Iterable, Iterable, List, list}
```

Comment:
I'm a bit conflicted on using `Iterable` here, mostly because of the need to move it to the end - is this really something we should be handling?  It's much more obvious to me to accept `Sequence` than `Iterable` (see also https://github.com/allenai/allennlp/issues/3562).

Reply:
(Also, after resolving this issue one way or the other, I think this looks good and is ready to merge; thanks for the PR!)

Reply:
> I'm a bit conflicted on using `Iterable` here, mostly because of the need to move it to the end - is this really something we should be handling? It's much more obvious to me to accept `Sequence` than `Iterable` (see also #3562).

Why? I think the author of #3562 suggests `Sequence` instead of `Iterable` there because `split_sizes` within the function `block_orthogonal` gets iterated many times. `Iterable` is the most general way of defining something you can iterate on.

We would still need to move it to the end even with `Sequence`. Could you explain why is this a problem?

Reply:
(`Sequence` is gonna be supported automatically by supporting `Iterable`)

Reply:
(reference: https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes)

Reply:
I think the distinction is that `Sequence` is not a super-type of `dict`, while `Iterable` is.  I don't think you would have to move this to the end if you just use `Sequence` (though you might need to move it after `Tuple`).  It's not the moving that bothers me, it's what the fact that you have to move it implies.  `Iterable` is a super general class; should our default behavior really be to create a list for it?  I'm more tempted to just punt on that case, and not try to construct something so general.  Though I can see an argument that constructing a list when you ask for an iterable isn't such a bad thing to do.  It just makes me pause, where adding `Sequence` here wouldn't make me pause at all.

Reply:
I think it makes sense for our default behavior (when many if-conditions aren't met) to create a list for it (or any iterable) provided it's `Iterable`.

Also, if we don't support it, users could use `Iterable` in their constructors and it'd fail.

Reply:
Well, yes, the question is whether it's reasonable to fail if someone chooses to use `Iterable` - what exactly did they want, and is a list sufficient?  It probably is.

Maybe just change the comment to say something like, "for any other kind of iterable, we will just assume that a list is good enough, and treat it that same as `List`.  This condition needs to be at the end, so we don't catch other kinds of `Iterables` with this branch."